compiler/rustc_ast/src/util/comments.rs

```diff
@@ -47,54 +46,66 @@ pub fn beautify_doc_string(data: Symbol) -> String {
             j -= 1;
         }
 
+        if i != 0 || j != lines.len() { Some((i, j)) } else { None }
+    }
+
+    /// remove whitespace-only lines from the start/end of lines
+    fn vertical_trim(lines: Vec<&str>, i: usize, j: usize) -> Vec<&str> {
         lines[i..j].to_vec()
     }
 
-    /// remove a "[ \t]*\*" block from each line, if possible
-    fn horizontal_trim(lines: Vec<String>) -> Vec<String> {
+    fn get_horizontal_trim(lines: &[&str]) -> Option<usize> {
         let mut i = usize::MAX;
-        let mut can_trim = true;
         let mut first = true;
 
-        for line in &lines {
+        for line in lines {
             for (j, c) in line.chars().enumerate() {
                 if j > i || !"* \t".contains(c) {
-                    can_trim = false;
-                    break;
+                    return None;
                 }
                 if c == '*' {
                     if first {
                         i = j;
                         first = false;
                     } else if i != j {
-                        can_trim = false;
+                        return None;
                     }
                     break;
                 }
             }
             if i >= line.len() {
-                can_trim = false;
-            }
-            if !can_trim {
-                break;
+                return None;
             }
         }
+        Some(i)
+    }
 
-        if can_trim {
-            lines.iter().map(|line| (&line[i + 1..line.len()]).to_string()).collect()
-        } else {
-            lines
-        }
+    /// remove a "[ \t]*\*" block from each line, if possible
+    fn horizontal_trim(lines: Vec<&str>, i: usize) -> Vec<String> {
+        lines.iter().map(|line| (&line[i + 1..line.len()]).to_string()).collect()
```

Comment:
Does this compile?
```suggestion
    fn horizontal_trim(lines: Vec<&str>, i: usize) -> Vec<&str> {
        for line in lines.iter_mut() {
            *line = &line[i+1..line.len()];
        }
        lines
```
If so that reduces both the second `Vec` allocation and re-allocating each line within the vec. It would be nice to change it to take `&mut [&str]` in that case.

Reply:
Apparently it's working! \o/